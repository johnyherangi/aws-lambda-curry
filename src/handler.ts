import { Context } from "aws-lambda"

export type Handler<TEvent, TResult> = (
    event: TEvent,
    context: Context,
) => TResult | Promise<TResult>
