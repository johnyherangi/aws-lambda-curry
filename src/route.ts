import { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from "aws-lambda"
import { Handler } from "./handler"
import { Middleware, middleware } from "./middleware"

export type HttpMethod =
    | "CONNECT"
    | "DELETE"
    | "GET"
    | "HEAD"
    | "OPTIONS"
    | "PATCH"
    | "POST"
    | "PUT"
    | "TRACE"

export type Route<TEvent, TResult> = [HttpMethod, string, Handler<TEvent, TResult>]

export function httpRoute(httpMethod: HttpMethod) {
    return function path(path: string) {
        return function handler(
            handler: Handler<APIGatewayProxyEvent, APIGatewayProxyResult>,
        ): Route<APIGatewayProxyEvent, APIGatewayProxyResult> {
            return [httpMethod, path, handler]
        }
    }
}

export function httpRoutes(routes: Route<APIGatewayProxyEvent, APIGatewayProxyResult>[]) {
    return function middlewareFn(
        middlewares?: Middleware<APIGatewayProxyEvent, APIGatewayProxyResult>[],
    ): Handler<APIGatewayProxyEvent, APIGatewayProxyResult> {
        const routeHandler = (event: APIGatewayProxyEvent, context: Context) => {
            const match = routes.find((route) => {
                const [httpMethod, path] = route
                return event.httpMethod === httpMethod && event.path === path
            })

            if (!match) {
                throw new Error("ROUTE_NOT_FOUND")
            }

            const [, , handler] = match
            return handler(event, context)
        }

        return middleware(middlewares)(routeHandler)
    }
}
