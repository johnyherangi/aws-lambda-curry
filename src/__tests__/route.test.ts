import { Handler } from "@src/handler"
import { httpRoute, httpRoutes } from "@src/route"
import { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from "aws-lambda"
import { Middleware } from ".."

describe("route.ts", () => {
    describe("httpRoute()", () => {
        it("creates http route", () => {
            const handler1: Handler<APIGatewayProxyEvent, APIGatewayProxyResult> = (e) => {
                return {
                    statusCode: 200,
                    body: "",
                }
            }

            const route = httpRoute("GET")("/example")(handler1)
            const [httpMethod, path] = route

            expect(httpMethod).toBe("GET")
            expect(path).toBe("/example")
        })
    })
    describe("httpRoutes()", () => {
        it("creates http route handler", () => {
            const handler1: Handler<APIGatewayProxyEvent, APIGatewayProxyResult> = (e) => {
                return {
                    statusCode: 200,
                    body: "",
                }
            }

            const middleware1: Middleware<APIGatewayProxyEvent, APIGatewayProxyResult> = (
                e,
                c,
                next,
            ) => next(e, c)

            const route1 = httpRoute("GET")("/example")(handler1)
            const handler = httpRoutes([route1])([middleware1])

            const event = {
                httpMethod: "GET",
                path: "/example",
            } as APIGatewayProxyEvent

            expect(handler(event, {} as Context)).toMatchSnapshot()
        })
        it("throws ROUTE_NOT_FOUND error", () => {
            const handler = httpRoutes([])()

            const event = {
                httpMethod: "GET",
                path: "/example",
            } as APIGatewayProxyEvent

            expect(() => handler(event, {} as Context)).toThrow("ROUTE_NOT_FOUND")
        })
    })
})
