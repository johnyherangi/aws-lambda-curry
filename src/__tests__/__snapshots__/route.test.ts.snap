// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route.ts httpRoutes() creates http route handler 1`] = `
{
  "body": "",
  "statusCode": 200,
}
`;
